### ISSUE TITLE
`.with()` replacement follow-up

### ISSUE THREAD

**Issue by finestructure** [SwiftPackageIndex/SwiftPackageIndex-Server#1278]
PR #1276 is the first of a number of changes to make queries with joins more efficient. This is an issue to track the follow-up work for other queries, marked as `TODO`s in code:

- [x] update `KeywordController.query`
- [x] remove `PackageInfo.init?(package:)`
- [x] 	review/update `ReAnalyzeVersions.swift`: `fetchReAnalysisCandidates(_:before:limit:)`
- [x] review/update `Package+ModelSupport.swift`: `Package.query(on:owner:repository:)`
- [x] 	review/update `Build+ModelSupport.swift`: `Build.query(on:)`

## Follow-up tasks triggered by the work on the above

- [x] query in `PackageCollection.generate` needs new query type
- [x] remove `Package.repository` (started on branch `remove-package-repository`, needs `PackageCollection.generate` change to distinct query type that joins `Version` with `Package` and references `Build`, `Product`, `Target`, `Repository`)
- [x] remove `Package.fetchRepository` (started on branch `remove-package-repository`)
- [x] remove `Version.fetchPackage` (started on branch `remove-package-repository`)
- [x] `Build.query` needs new query type
- [x] check TODO in `BuildShow+Model.swift` to replace `let version = build.$version.value`
- [x] check TODO in `JPRVB+ModelSupport.swift` to skip the check on `version.$reference.value`
- [x] add `JPR.query(on:owner:repository)` to allow narrowing of type in `PackageController.swift` (see [Joined2 refactor (part 2) by finestructure · Pull Request #1292 · SwiftPackageIndex/SwiftPackageIndex-Server · GitHub](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/pull/1292#discussion_r727152293))

**Comment by finestructure** [comment]
Changing this for `ReAnalyzeVersions` is a worthwhile but bigger undertaking. I'm outlining the changes here to be picked up in the future because it may not be the right time to do this (more on that below).

- `ReAnalyzeVersions` should use a join over a `.with` but it's a 2-way-join, not a three way. I've made `JoinedQueryBuilder` generic, renamed `Joined<M,R1,R2>` to `Joined3<M,R1,R2>`, and added `Joined2<M,R>` to prepare for this.
- it's not just `ReAnalyzeVersions` that uses the pattern `Package.query(...).with(\.$repositories)`. It's also
  - `Package.fetchCandidate`
  - `Package.fetchCandidates`
  
  and all three feed into the same analysis and ingestion methods. I.e. changing even one has a knock-on effect of requiring all subsequent calls and their callers to change signature from `Package` to `Joined2<Package, Repository>`
- This _should_ be a fairly mechanical change with the compiler guiding us through but it will touch all central parts of processing
- The big gain here is that what's currently purely a "test asserted" reliance on the initial query pulling in the repository data would be encoded in the type - and it would save an additional query roundtrip, of course.

Now the reason we probably shouldn't be doing this right now is that I suspect we'll see significant simplification to the processing once we introduce async/await. In particular, I expect a few of the overloads we've introduced to make the "map stacks" more readable to go away entirely.

For instance, we have a common pattern of "array handler" calling "single instance hander", say `updateVersions` wrapping `updateVersion`. This deals with continuation in case of one "single instance handler" throwing an error.

I expect all of these to collapse into a shorter single function with async/await because it'll remove the mechanics of `EventLoopFuture.andAllComplete` and the like.

It should be much easier to make the change from `Package` → `Joined2<Package, Repository>` after this change.

I'll spend a little more time exploring this and at the very least I'll prepare a PR with the generic `Joined2` changes for future use.
